src/Sarif.FunctionalTests/v2/ConverterTestData/PREfast/Expected-022.xml.sarif

```diff
@@ -12,46 +12,55 @@
           "mimeType": "text/x-cpp"
         }
       },
-      "logicalLocations": {
-        "f": {
+      "logicalLocations": [
+        {
+          "fullyQualifiedName": "f",
           "decoratedName": "?f@@YAXXZ"
         },
-        "h": {
+        {
+          "fullyQualifiedName": "h",
           "decoratedName": "?h@@YAXPAHH@Z"
         },
-        "foo": {
+        {
+          "fullyQualifiedName": "foo",
           "decoratedName": "?foo@@YAXPAHI@Z"
         },
-        "StrLen": {
+        {
+          "fullyQualifiedName": "StrLen",
           "decoratedName": "?StrLen@@YAIPBD@Z"
         },
-        "BugLast": {
+        {
+          "fullyQualifiedName": "BugLast",
           "decoratedName": "?BugLast@@YADPBD@Z"
         },
-        "g1": {
+        {
+          "fullyQualifiedName": "g1",
           "decoratedName": "?g1@@YAXPAH@Z"
         },
-        "foo-0": {
-          "name": "foo",
+        {
           "fullyQualifiedName": "foo",
           "decoratedName": "?foo@@YADXZ"
         },
-        "bar": {
+        {
+          "fullyQualifiedName": "bar",
           "decoratedName": "?bar@@YAXPAHII@Z"
         },
-        "ShiftTest::f": {
+        {
           "name": "f",
```

Comment:
>, [](start = 21, length = 1)

So yes, you did the right thing by supplying both `name` and `fQLN` for non-top-level locations. It's just that for top-level locations, it's `name` that's required, and `fQLN` is optional.